src/orchestrator.ts

```diff
@@ -550,28 +542,28 @@ export class Orchestrator {
         }
     }
 
-    private all(state: HistoryEvent[], tasks: Task[]): TaskSet {
-        const faulted = tasks.filter(TaskFilter.isFailedTask);
-        if (faulted.length > 0) {
-            return TaskFactory.FailedTaskSet(tasks, (faulted[0] as FailedSingleTask).exception);
-        }
-
-        const succesfulTasks = tasks.filter(TaskFilter.isSuccessfulSingleTask);
-        const isCompleted = succesfulTasks.length === tasks.length;
-        if (isCompleted) {
-            const results = succesfulTasks.reduce((acc, t) => {
-                return [...acc, t.result];
-            }, []);
+    private all(state: HistoryEvent[], tasks: TaskBase[]): TaskSet {
+        const completedTasks = tasks.filter(TaskFilter.isCompletedTask);
+        if (completedTasks.length === tasks.length) {
+            const sortedTasks = completedTasks.sort(TaskFilter.CompareFinishedTime);
+            const completionIndex = sortedTasks[tasks.length - 1].completionIndex;
 
-            return TaskFactory.SuccessfulTaskSet(tasks, results);
+            const failedTasks = completedTasks.filter(TaskFilter.isFailedTask);
+            if (failedTasks.length > 0) {
+                // TODO: aggregate all failures into one clean error.
+                return TaskFactory.FailedTaskSet(tasks, completionIndex, (failedTasks[0] as FailedTask).exception);
+            } else {
+                const results = completedTasks.map((task) => task.result);
+                return TaskFactory.SuccessfulTaskSet(tasks, completionIndex, results);
+            }
         } else {
             return TaskFactory.UncompletedTaskSet(tasks);
         }
     }
 
-    private any(state: HistoryEvent[], tasks: Task[]): TaskSet {
+    private any(state: HistoryEvent[], tasks: TaskBase[]): TaskSet {
         const completedTasks = tasks
-            .filter(TaskFilter.isSuccessfulSingleTask)
+            .filter(TaskFilter.isCompletedTask)
             .sort((a, b) => {
```

Comment:
It looks like here and also in the `all` method we're sorting just to find the min or max element. If we don't actually need a sorted list, it would be more CPU efficient to just get the min or max of the list. It looks like this should be fairly straightforward to do in ES6: https://stackoverflow.com/a/59511498/2069